### ISSUE TITLE
Proposal: To enhance RuleTester to detect whether or not AST was modified.

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#4156]
Recently I saw a bug related to a rule which modifies AST node: https://github.com/yannickcr/eslint-plugin-react/issues/249
So I'd like to revisit #1231.

My idea is to enhance RuleTester to detect whether or not AST was modified.

1. It clones AST node deeply (excludes the `parent` property) and saves that before it runs a test case.
2. It clones AST node deeply (excludes the `parent` property) and saves that after it runs a test case.
3. `assert.deepEqual(beforeAST, afterAST)`.

----

I'm happy to work on this if accepted.

**Comment by gyandeeps** [comment]
I like the idea.

**Comment by ilyavolodin** [comment]
I think it's a good idea, but I think it might significantly slow down our tests. And they are not fast to begin with.

**Comment by platinumazure** [comment]
Could we hash the AST nodes somehow and compare that?

**Comment by mysticatea** [comment]
In my PC, `npm test` became `43s` from `36s`.
Yes, it slowed down.

I thought information of mocha's `assert.deepEqual` is helpful.

```diff
AssertionError: Should not modify AST
+ expected - actual

                 "column": 4
                 "line": 1
               }
             }
-            "name": "foo!"
+            "name": "foo"
             "range": [
               4
               7
             ]
```

**Comment by nzakas** [comment]
We could just run this on the AST before each test: https://github.com/substack/deep-freeze

Then an error will be thrown on attempts to modify the AST. We would need to check performance, but I think it's the simplest option.

**Comment by mysticatea** [comment]
I guess we cannot use `deep-freeze`, because eslint add `parent` property while traversing.
But, "Then an error will be thrown on attempts to modify the AST." this is great.
I will try to use `Object.preventExtensions()` and `Object.defineProperty()`.

**Comment by mysticatea** [comment]
I tried to freeze AST (excludes `parent`).
I could get exact stack traces, yay.
But it seems to be slower than `assert.deepEqual` version (I'm not sure why).

```js
function freezeDeeplyExcludesParent(x) {
    var i;

    if (typeof x === "object" && x !== null) {
        if (Array.isArray(x)) {
            Object.freeze(x);

            for (i = 0; i < x.length; ++i) {
                freezeDeeplyExcludesParent(x[i]);
            }
        } else {
            var keys = Object.keys(x);
            for (i = 0; i < keys.length; ++i) {
                var key = keys[i];
                if (key === "parent") {
                    continue;
                }

                Object.defineProperty(x, key, {
                    value: freezeDeeplyExcludesParent(x[key]),
                    configurable: false,
                    enumerable: true,
                    writable: false
                });
            }

            // Reserves `parent` property and prevent extensions.
            Object.defineProperty(x, "parent", {
                value: x.parent,
                configurable: false,
                enumerable: true,
                writable: true
            });
            Object.preventExtensions(x);
        }
    }

    return x;
}
```

**Comment by nzakas** [comment]
Hmmm. Okay, `assert.deepEqual` it is :)
